src/utils/crypto.ts

```diff
@@ -0,0 +1,135 @@
+import type { IKey } from '@/types';
+import { STORAGE_KEYS, PASSWORD_ENCRYPTION_ALGO, IS_EXTENSION } from '@/constants';
+import { WalletStorage } from '@/lib/WalletStorage';
+import { toRaw } from 'vue';
+
+/**
+ * If the application is running as an extension, the key should be extractable.
+ * So that we can save it in the session storage.
+ */
+const IS_EXTRACTABLE = !!IS_EXTENSION;
+
+// Utility Functions for Base64 Encoding and Decoding
+const encodeBase64 = (data: Uint8Array) => Buffer.from(data).toString('base64');
+const decodeBase64 = (data: string) => Buffer.from(data, 'base64');
+
+export async function generateKey(password: string, encryptedMnemonic?: string): Promise<IKey> {
+  const subtleCrypto = globalThis.crypto.subtle;
+  const data = encryptedMnemonic
+    ? new Uint8Array(decodeBase64(encryptedMnemonic))
+    : null;
+
+  // Generate a random salt or extract it from the data
+  const salt = data ? data.slice(0, 16) : globalThis.crypto.getRandomValues(new Uint8Array(16));
+
+  // Generate a random IV or extract it from the data
+  const iv = data ? data.slice(16, 32) : globalThis.crypto.getRandomValues(new Uint8Array(16));
+
+  // Encode password to ArrayBuffer
+  const enc = new TextEncoder();
+  const passwordBuffer = enc.encode(password);
+
+  // Derive the key using PBKDF2
+  const keyMaterial = await subtleCrypto.importKey(
+    'raw',
+    passwordBuffer,
+    { name: 'PBKDF2' },
+    false,
+    ['deriveKey'],
+  );
+
+  const key = await subtleCrypto.deriveKey(
+    {
+      name: 'PBKDF2',
+      salt,
+      iterations: 100000,
+      hash: 'SHA-256',
+    },
+    keyMaterial,
+    { name: PASSWORD_ENCRYPTION_ALGO, length: 256 },
+    IS_EXTRACTABLE,
+    ['encrypt', 'decrypt'],
+  );
+
+  return { key, salt, iv };
+}
+
+export async function encrypt(passwordKey: IKey, plaintext: string): Promise<string> {
+  const { key, salt, iv } = toRaw(passwordKey);
```

Comment:
Are you sure the `toRaw` is needed here? Using this means that the `passwordKey` could be something else than `IKey`.